### ISSUE TITLE
pipelined crash observed with HIL sanity run

### ISSUE THREAD

**Issue by milapjoshi** [magma/magma#10746]
### Your Environment

- **Version:** 1.7.0-1638987121-f05b82db
- **Affected Component:** Access Gateway
- **Affected Subcomponent:**AGW service
- **Deployment Environment:** bare-metal (AGW)

### Describe the Issue

During TC009 and TC010 run we have observed that sessiond failed to activate all rule for IMSI and send termination 

```
Dec  8 11:50:29 hil-1 sessiond[3411]: I1208 11:50:29.328683  3426 LocalEnforcer.cpp:1971] Could not activate rules for IMSI001011234560533-725618, rpc failed: ActivateFlows processing timed out, terminating session...
Dec  8 11:50:29 hil-1 sessiond[3411]: I1208 11:50:29.330199  3411 LocalEnforcer.cpp:561] Initiating session termination for IMSI001011234560533-725618
Dec  8 11:50:29 hil-1 sessiond[3411]: I1208 11:50:29.330559  3411 SpgwServiceClient.cpp:66] Deleting default bearer and corresponding PDN session for IMSI: IMSI001011234560533 APN IP addr 192.168.130.179 Bearer ID 5
```

however while removing QoS flows pipelined crash with below stacktrace

```
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.common:removing Qos for imsi IMSI001011234560526 del_rule_num 527
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.common:removing rule 001011234560526 527
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.qos_tc_impl:deleting qos_handle 2105, skip_filter False
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:Del Filter iface eth0 qid 0x839
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.tc_pyroute2:del-filter error : 2
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: Traceback (most recent call last):
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/qos/tc_ops_pyroute2.py", line 135, in del_filter
Dec  8 11:50:29 hil-1 pipelined[3489]:     ret = self._ipr.tc(
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/iproute/linux.py", line 1614, in tc
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self.nlm_request(msg, msg_type=command, msg_flags=flags))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 376, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_request(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 867, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     for msg in self.get(msg_seq=msg_seq,
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 379, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_get(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 704, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     raise msg['header']['error']
Dec  8 11:50:29 hil-1 pipelined[3489]: pyroute2.netlink.exceptions.NetlinkError: (2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:Delete HTB iface eth0 qid 0x839
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.tc_pyroute2:del-htb  error error : 2
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: Traceback (most recent call last):
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/qos/tc_ops_pyroute2.py", line 91, in del_htb
Dec  8 11:50:29 hil-1 pipelined[3489]:     ret = self._ipr.tc("del-class", "htb", if_index, htb_queue)
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/iproute/linux.py", line 1614, in tc
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self.nlm_request(msg, msg_type=command, msg_flags=flags))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 376, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_request(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 867, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     for msg in self.get(msg_seq=msg_seq,
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 379, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_get(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 704, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     raise msg['header']['error']
Dec  8 11:50:29 hil-1 pipelined[3489]: pyroute2.netlink.exceptions.NetlinkError: (2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.qos_tc_impl:error deleting class 2105, not releasing idx
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.qos_tc_impl:deleting qos_handle 2104, skip_filter False
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:Del Filter iface eth1 qid 0x838
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.tc_pyroute2:del-filter error : 2
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: Traceback (most recent call last):
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/qos/tc_ops_pyroute2.py", line 135, in del_filter
Dec  8 11:50:29 hil-1 pipelined[3489]:     ret = self._ipr.tc(
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/iproute/linux.py", line 1614, in tc
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self.nlm_request(msg, msg_type=command, msg_flags=flags))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 376, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_request(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 867, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     for msg in self.get(msg_seq=msg_seq,
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 379, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_get(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 704, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     raise msg['header']['error']
Dec  8 11:50:29 hil-1 pipelined[3489]: pyroute2.netlink.exceptions.NetlinkError: (2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:Delete HTB iface eth1 qid 0x838
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.tc_pyroute2:del-htb  error error : 2
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: Traceback (most recent call last):
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/qos/tc_ops_pyroute2.py", line 91, in del_htb
Dec  8 11:50:29 hil-1 pipelined[3489]:     ret = self._ipr.tc("del-class", "htb", if_index, htb_queue)
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/iproute/linux.py", line 1614, in tc
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self.nlm_request(msg, msg_type=command, msg_flags=flags))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 376, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_request(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 867, in nlm_request
Dec  8 11:50:29 hil-1 pipelined[3489]:     for msg in self.get(msg_seq=msg_seq,
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 379, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     return tuple(self._genlm_get(*argv, **kwarg))
Dec  8 11:50:29 hil-1 pipelined[3489]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 704, in get
Dec  8 11:50:29 hil-1 pipelined[3489]:     raise msg['header']['error']
Dec  8 11:50:29 hil-1 pipelined[3489]: pyroute2.netlink.exceptions.NetlinkError: (2, 'No such file or directory')
Dec  8 11:50:29 hil-1 pipelined[3489]: ERROR:pipelined.qos.qos_tc_impl:error deleting class 2104, not releasing idx
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.common:removing root ambr qos handle 2103 direction 0
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.qos_tc_impl:deleting qos_handle 2103, skip_filter True
```
ii  python3-pyroute2                     0.5.14                                all          Python Netlink library
hil-1/ and hil-2 is SUT

**Comment by pshelar** [comment]
Looks like the python file is missing:
```
Dec  8 11:50:29 hil-1 pipelined[3489]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
```

@milapjoshi can you check if 'python3-pyroute2' package is installed on AGW?

**Comment by uri200** [comment]
Hey @pshelar , per @milapjoshi  comment the version is 
```
ii python3-pyroute2 0.5.14 all Python Netlink library
hil-1/ and hil-2 is SUT
```

**Comment by pshelar** [comment]
actually ignore my prev comment.
Its trying to delete non existing TC object, This is due to session timeout exception:
```
Dec  7 06:35:29 hil-1 mme[3520]: [DEBUG] Success logging event: detach_success
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:adding qos for imsi IMSI001011234560515 rule_num 516 direction 0 apn_ambr 100000000, None
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:existing root rec: ambr_qos_handle_root 0
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:add QoS: QosInfo(gbr=None, mbr=100000000)
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.id_manager:allocating idx 2063
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:assigned qid: 2063
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Added root ambr qos mbr 100000000 direction 0 qos_handle 2063
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:existing leaf rec: ambr_qos_handle_leaf 0
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:add QoS: QosInfo(gbr=None, mbr=100000000)
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.id_manager:allocating idx 2065
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:assigned qid: 2065
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Added ambr qos mbr 100000000 direction 0 qos_handle 2063/2065
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Update: ["Subscriber", "001011234560515", "192.168.133.84", 516, 0] -> ["qos_data", 2065, 2063, 2065]
Dec  7 06:35:29 hil-1 pipelined[3481]: ERROR:root:ActivateFlows request processing timed out
>>>> Dec  7 06:35:29 hil-1 sessiond[3410]: I1207 06:35:29.213553  3448 LocalEnforcer.cpp:1966] Could not activate rules for IMSI001011234560515-984070, rpc failed: ActivateFlows processing timed out, terminating session...
Dec  7 06:35:29 hil-1 mobilityd[3415]: INFO:root:Released IP 192.168.141.55 for sid IMSI001011234560524
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:EnforcementController:adding Actions OFPActionSetField(pkt_mark=2065) instruction None
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:adding qos for imsi IMSI001011234560515 rule_num 516 direction 1 apn_ambr 100000000, None
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:existing root rec: ambr_qos_handle_root 0
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:add QoS: QosInfo(gbr=None, mbr=100000000)
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.id_manager:allocating idx 2062
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:assigned qid: 2062
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Added root ambr qos mbr 100000000 direction 1 qos_handle 2062
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:existing leaf rec: ambr_qos_handle_leaf 0
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:add QoS: QosInfo(gbr=None, mbr=100000000)
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.id_manager:allocating idx 2064
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:assigned qid: 2064
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Added ambr qos mbr 100000000 direction 1 qos_handle 2062/2064
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:Update: ["Subscriber", "001011234560515", "192.168.133.84", 516, 1] -> ["qos_data", 2064, 2062, 2064]
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:EnforcementController:adding Actions OFPActionSetField(pkt_mark=2064) instruction None
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:ryu.base.app_manager:EVENT ofp_event->InOutController EventOFPBarrierReply
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:ryu.base.app_manager:EVENT ofp_event->GYController EventOFPBarrierReply
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:ryu.base.app_manager:EVENT ofp_event->EnforcementController EventOFPBarrierReply
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:ryu.base.app_manager:EVENT ofp_event->EnforcementStatsController EventOFPBarrierReply
Dec  7 06:35:29 hil-1 pipelined[3481]: DEBUG:ryu.base.app_manager:EVENT ofp_event->ofctl_service EventOFPBarrierReply
```
later on session cleanup results in error:
```
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:removing Qos for imsi IMSI001011234560515 del_rule_num 516
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.common:removing rule 001011234560515 516
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:deleting qos_handle 2065, skip_filter False
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:Del Filter iface eth0 qid 0x811
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:Return: ({'header': {'length': 36, 'type': 2, 'flags': 256, 'sequence_number': 441, 'pid': 3481, 'error': None, 'target': 'localhost', 'stats': Stats(qsize=0, delta=0, delay=0)}, 'event': 'NLMSG_ERROR'},)
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:Delete HTB iface eth0 qid 0x811
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:Return: ({'header': {'length': 36, 'type': 2, 'flags': 256, 'sequence_number': 442, 'pid': 3481, 'error': None, 'target': 'localhost', 'stats': Stats(qsize=0, delta=0, delay=0)}, 'event': 'NLMSG_ERROR'},)
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.id_manager:releasing idx 2065
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.qos_tc_impl:deleting qos_handle 2064, skip_filter False
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:Del Filter iface eth1 qid 0x810
Dec  7 06:35:31 hil-1 pipelined[3481]: ERROR:pipelined.qos.tc_pyroute2:del-filter error : 2
Dec  7 06:35:31 hil-1 pipelined[3481]: DEBUG:pipelined.qos.tc_pyroute2:(2, 'No such file or directory')
Dec  7 06:35:31 hil-1 pipelined[3481]: Traceback (most recent call last):
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/qos/tc_ops_pyroute2.py", line 135, in del_filter
Dec  7 06:35:31 hil-1 pipelined[3481]:     ret = self._ipr.tc(
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/iproute/linux.py", line 1614, in tc
Dec  7 06:35:31 hil-1 pipelined[3481]:     return tuple(self.nlm_request(msg, msg_type=command, msg_flags=flags))
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 376, in nlm_request
Dec  7 06:35:31 hil-1 pipelined[3481]:     return tuple(self._genlm_request(*argv, **kwarg))
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 867, in nlm_request
Dec  7 06:35:31 hil-1 pipelined[3481]:     for msg in self.get(msg_seq=msg_seq,
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 379, in get
Dec  7 06:35:31 hil-1 pipelined[3481]:     return tuple(self._genlm_get(*argv, **kwarg))
Dec  7 06:35:31 hil-1 pipelined[3481]:   File "/usr/local/lib/python3.8/dist-packages/pyroute2/netlink/nlsocket.py", line 704, in get
Dec  7 06:35:31 hil-1 pipelined[3481]:     raise msg['header']['error']
Dec  7 06:35:31 hil-1 pipelined[3481]: pyroute2.netlink.exceptions.NetlinkError: (2, 'No such file or directory')
```
This is already handled in the code, we can improve log msg for such cases. so I will keep this issue open.

**Comment by pshelar** [comment]
Just to be clear, this is not crashing pipelineD, the exception is logged and handled gracefully.
